File: C/Stack.c
#include <stdio.h>

/* Define a 'Node' struct, which includes an integer value and a pointer
   to the next node. */
typedef struct Node
{
	int value;
	struct Node* next;
} Node;

// Add nodes to the top of the stack.
void push(Node** top, int val)
{
	// Create new node
	Node* newTop = malloc(sizeof(Node));

	// Set new node's value and link to previous top node
	newTop->value = val;
	newTop->next = *top;
	*top = newTop;
}

// Remove nodes from the top of the stack (and return their values).
int pop(Node** top)
{
	// Check for empty stack
	if (*top == NULL)
	{
		return NULL;
	}

	// Save top node's parameters, then delete it.
	int val = (*top)->value;
	Node* newTop = (*top)->next;
	free(*top);

	// Reassign top node pointer and return saved value.
	*top = newTop;
	return val;
}

// Prints a stack in a readable format.
void printStack(Node* top)
{
	Node* cur = top;
	
	// Check for empty stack
	if (top == NULL)
	{
		printf("Stack(");
	}

	// Print each node's value until the end of the stack is reached.
	else
	{
		printf("Stack(%d", cur->value);
		while (cur->next != NULL)
		{
			cur = cur->next;
			printf(", %d", cur->value);
		}
	}
	
	// End of stack
	printf(")\n");
}

int main()
{
	// Initialize stack and test push function
	printStack(NULL);
	Node* top = malloc(sizeof(Node));
	top->value = 3;
	top->next = NULL;
	push(&top, 21);
	push(&top, 65);
	printStack(top);

	// Test pop function
	int a = pop(&top);
	printStack(top);
	int b = pop(&top);
	printStack(top);
	int c = pop(&top);
	printStack(top);

	return 0;
}

File: C#/Stack.cs
using System;

namespace Structure
{
    // Nodes contain a value and reference to the next node.
    class Node
    {
        public int value;
        public Node next;

        public Node(int val)
        {
            value = val;
        }
    }

    // Implement a stack (FIFO) structure.
    class Stack
    {
        public Node head;

        // Insert a new node on the top of the stack.
        public void push(int val)
        {
            // Check for empty stack
            if (head == null)
            {
                head = new Node(val);
            }

            // Add new node and have it reference the old head node
            else
            {
                Node prvHead = head;
                head = new Node(val);
                head.next = prvHead;
            }
        }

        // Remove a node from the top of the stack. Return its value.
        public int pop()
        {
            if (head != null)
            {
                int returnVal = head.value;
                head = head.next;
                return returnVal;
            }

            // Throw exception if stack is empty.
            throw new InvalidOperationException("Empty Stack");
        }

        // Print the stack in a readable format.
        public void prtStack()
        {
            // Check for empty stack
            if (head == null)
            {
                Console.WriteLine("Stack()");
            }

            // Print each node of stack.
            else
            {
                Node cur = head;
                Console.Write("Stack({0}", cur.value);
                while (cur.next != null)
                {
                    cur = cur.next;
                    Console.Write(", {0}", cur.value);
                }
                Console.Write(")\n");
            }
        }
    }

    class Program
	{
		static int Main()
		{
            // Initialize stack and test push function
            Stack stk = new Stack();
            stk.push(83);
            stk.push(73);
            stk.push(8);
            stk.prtStack();

            // Test pop function
            int a = stk.pop();
            stk.prtStack();
            int b = stk.pop();
            stk.prtStack();
            int c = stk.pop();
            stk.prtStack();
            return 0;
		}
	}
}